### ISSUE TITLE
Significantly reduce the space required by user downloads

### ISSUE THREAD

**Issue by MattBlissett** [gbif/portal16#661]
This applies only to large downloads, e.g. over 50GB. A full DWCA download of everything is 180GB.

### Space used

0.4% of our downloads (1779) are over 50GB, they take up about half the total storage of the cluster (301TB with replicants, 100TB without).

The other 99.6%, the 426362 downloads under 50GB, consume just 27TB (with replicants).

If we reduce the storage used by large downloads, we can store the smaller downloads pretty much indefinitely.

### Costs

This is presently "free", in that the hard drives are needed for fast data storage and indexing of the occurrence store. But it's not sustainable at the current rate of large downloads, and buying additional hardware is unlikely to be economical.

With Amazon or Google archive storage, a 180GB everything-download would cost $9/year to store, and $20 to retrieve. Thus the 0.4% would cost $4800 per year to store. Copenhagen University's storage is available to us, and although it isn't archive storage, it is still competitive as there is no retrieval fee. (I don't know if the price is confidential, I won't write it here.)

That's reasonable, but I think it still means storing a lot of unwanted/duplicated data.

### Unwanted downloads

The logs show that about half the time, users download the entire occurrence dataset more than once within two days. Either there is a UI problem, or they are unable to open the ZIP/CSV and assume the download is corrupt, so repeat it. Excel can handle 1 million rows, MS Access 2GB of data.

Other users retrieve all the data, but have no intention of citing it — or they realise their download filter is wrong, and produce a second download and cite that instead.

### Proposals:

1. For large downloads, especially unfiltered downloads, provide advice on the download page. Does the user really want to download everything? Can their workflow handle it? Extra-especially if they already downloaded everything yesterday.
2. For large downloads that go ahead, default to deleting them after a certain period of time, e.g. 6 months. The user can extend this — possibly indefinitely, or maybe for fixed periods until we see a citation 😈, which would then keep the download forever.
3. Provide weekly unfiltered downloads, with appropriate metadata, and point users towards this. That would be about 50TB after 5 years.

① can be done immediately, ② requires some work in the registry, ③ a bit more work in the registry and occurrence download service.

### Not useful:

* Daily unfiltered downloads, or reusing a download where the filter is the same, does not save enough space.

### When?

This needs to be done before the end of February, or we will run out of space.

**Comment by dschigel** [comment]
Storing everything forever is not practical. I discussed with Guy Cochrane about two years ago compression approaches NCBI takes, which was supposed to be lossless, but there were issues too - can't remember the details. As discussed several times, the fact that a download is cited through DOI in a research paper or in the policy context should immunize the download from deletion, no matter how large. The question is than for how long? No access via a DOI for N years - delete data, keep DOI map? We may introduce size and time of storage limits, and we should be very open about that. Please note that DOI for downloads story develops really well, and is one of the most powerful engagement instruments we have. If downloads start to disappear and this will surprise users and readers, we may start losing loyal users and never see them back - red alert from Dmitry

**Comment by dhobern** [comment]
Hi Matt - I have been proposing for some time that we do all three things you suggest, but I would recommend being a little more aggressive than you indicate around bullets 2 and 3.

I think we should apply this "dead-man's-pedal" approach to ALL downloads for consistency.  Any cut-off is arbitrary and the reminder messages to users can serve as a trigger for them to register a DOI for any relevant publications, even if they did not cite properly.

I also think that fixed monthly snapshots would be plenty - they don't need to be weekly.

**Comment by MattBlissett** [comment]
Number 1 is now done and live on gbif.org, except (I think) taking into account of an earlier, identical download.

**Comment by MattBlissett** [comment]
I've added a new field to Download, called `eraseAfter`, and added `FILE_ERASED` to the status enumeration.

Examples:

An erase date in the future:
https://api.gbif-dev.org/v1/occurrence/download/0000101-180109225247373
https://www.gbif-dev.org/occurrence/download/0000101-180109225247373

An erase date in the past, and the file has been erased:
https://api.gbif-dev.org/v1/occurrence/download/0000097-180109225247373
https://www.gbif-dev.org/occurrence/download/0000097-180109225247373

There are already API calls to update download metadata, so I think the portal can already implement some UI for the user to extend the deletion date, or mark it to be kept forever.

Nothing more is happening behind the scenes. I don't know if we want auto-deletion when the expiry date is reached; not to begin with anyway. We will need some auto-emails sent out as the date approaches.

**Comment by MortenHofft** [comment]
When will the API to delete/postpone deletion be in prod and what is the calls we should do and how should it be signed?

![screen shot 2018-02-06 at 09 06 16](https://user-images.githubusercontent.com/6759267/35848127-3ad08b98-0b1d-11e8-815a-9dceebc90a2d.png)

**Comment by MortenHofft** [comment]
btw, the `FILE_ERASED` enum is wonderful. So far the website has done a `get headers` and used that to guess if the file has been deleted.

**Comment by MattBlissett** [comment]
The download endpoint already accepts authenticated PUT requests to update the records (e.g. a PUT to https://api.gbif-dev.org/v1/occurrence/download/0000093-180109225247373). Authentication should have the same possibilities as the download request API (the user, or a special GBIF key).

Previously deleted downloads will indeed be updated with `FILE_ERASED`.

For the moment, "Delete download" would just be setting eraseAfter to now(). Deletions will be a manual process, and I'll have about 5 months before the first automatic notification emails need to be sent.

"Postpone" can add some months, and I think the user should be able to ask to keep a download indefinitely. I'm not sure how best to catch a citation at the same time.

**Comment by MattBlissett** [comment]
I will look at getting this into UAT this week. Deployment to prod ought to be coordinated, as the updated "Your GBIF download is ready" email has text saying an extension can be requested.

**Comment by MattBlissett** [comment]
This is now on UAT.

Example: https://www.gbif-uat.org/occurrence/download/0000012-180202174727027

I've also set the status to FILE_ERASED for all downloads which don't exist on UAT, which is all but three as the registry is a fresh copy from prod.

Example: https://www.gbif-uat.org/occurrence/download/0021422-160910150852091

**Comment by MattBlissett** [comment]
What's done is now in production:
- downloads are marked for deletion 6 months after creation.
- the user can extend this time by an additional six months (actually, a year after today whenever they request)
- deleted download files show as such

Not yet implemented:
- mailing users as the date approaches
- actually deleting files
or pre-made monthly downloads.

These will wait a little longer.

**Comment by MortenHofft** [comment]
> downloads are marked for deletion 6 months after creation.

I still don't see this in production. That is: there is no eraseAfter date in the response of fx http://api.gbif.org/v1/occurrence/download/0001558-171109162308116

**Comment by MattBlissett** [comment]
Yes, only on new ones.

I'll set it on old ones once the email reminders are ready.
